packages/@lwc/template-compiler/src/parser/expression-complex/html.ts

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { parseExpressionAt } from 'acorn';
+import { Document, DocumentFragment, ParsingError, ParserOptions } from 'parse5';
+import { ParserDiagnostics, invariant } from '@lwc/errors';
+import Parser from 'parse5/lib/parser';
+import Tokenizer from 'parse5/lib/tokenizer';
+import { TMPL_EXPR_ECMASCRIPT_EDITION } from '../constants';
+import type ParserCtx from '../parser';
+import type { PreparsedExpressionMap } from './types';
+
+const OPENING_CURLY_BRACKET = 0x7b;
+const CLOSING_CURLY_BRACKET = 0x7d;
+const WHITESPACE = /\s*/;
+const TRAILING_SPACES_AND_PARENS = /[\s)]*/;
+
+function getWhitespaceLen(str: string): number {
+    return WHITESPACE.exec(str)![0].length;
+}
+
+function getTrailingChars(str: string): string {
+    return TRAILING_SPACES_AND_PARENS.exec(str)![0];
+}
+
+function validateMatchingExtraParens(leadingChars: string, trailingChars: string) {
+    const numLeadingParens = leadingChars.split('(').length - 1;
+    const numTrailingParens = trailingChars.split(')').length - 1;
+    invariant(
+        numLeadingParens === numTrailingParens,
+        ParserDiagnostics.TEMPLATE_EXPRESSION_PARSING_ERROR,
+        ['expression must have balanced parentheses.']
+    );
+}
```

Comment:
Worth documenting: we don't need to handle something like `(foo())` ("unbalanced') ourselves because Acorn will consume the function-call's `)`, but not the final `)`.

Reply:
I'll add a comment here, as well as where the parenthesis-handling is done in the lexer.